purchase_invoice_new_picking_line/models/purchase_order.py

```diff
@@ -0,0 +1,81 @@
+# SPDX-FileCopyrightText: 2024 Coop IT Easy
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+from odoo import api, fields, models
+from odoo.fields import Command
+from odoo.tools.float_utils import float_compare, float_round
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    def _prepare_account_move_lines_from_stock_move(
+        self, stock_move_id, account_move_id
+    ):
+        self.ensure_one()
+
+        qty = max(stock_move_id.quantity_done, 0)
+        date = stock_move_id.date or fields.Date.today()
+        supplierinfo = self.env["product.supplierinfo"]
+        price_unit = 0.0
+        if self.partner_id:
+            supplierinfo = stock_move_id.product_id._select_seller(
+                partner_id=self.partner_id,
+                quantity=qty,
+                date=date,
+                uom_id=stock_move_id.product_uom,
+                params={"order_id": self},
+            )
+        if supplierinfo:
+            price_unit = supplierinfo.price
+        else:
+            price_unit = stock_move_id.product_id.lst_price
+        taxes = stock_move_id.product_id.supplier_taxes_id.filtered(
+            lambda tax: tax.company_id == self.company_id
+        )
+        tax_ids = self.fiscal_position_id.map_tax(taxes)  # TODO not sure about this
+
+        data = {
+            "display_type": "product",
+            # Normally the name is '{order_name}: {order_line_name}'.
+            "name": "%s: %s" % (stock_move_id.picking_id.name, stock_move_id.name),
+            "product_id": stock_move_id.product_id.id,
+            "product_uom_id": stock_move_id.product_uom.id,
+            "quantity": qty,
+            "price_unit": self.currency_id._convert(
+                price_unit,
+                self.company_id.currency_id,
+                self.company_id,
+                date,
+                round=False,
+            ),
+            "tax_ids": [Command.set(tax_ids.ids)],
+        }
+
+        return data
+
+    def add_missing_picking_products_to_account_move(self, account_move_id):
```

Comment:
This could theoretically be called from `purchase_order_change` in `account.move`, which has changed in semantics since Odoo 12. `purchase_order` is not stored on `account.move`. Instead, if you add a purchase order in the interface, it 'loads' data from the purchase order. I don't know if `purchase_order_change` is idempotent—this method is not.